### PR TITLE
Script which can convert existing bots in the group collector to Systemd services

### PR DESCRIPTION
**Pull request by navtej**
Title says it all. However this is a hackish script. For those who know what they are doing.

### CODE REVIEW

File: contrib/systemd/conf.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import shutil
+
+INTELMQ_DIR = '/opt/intelmq'
+RUNTIME_CONF = INTELMQ_DIR+'/etc/runtime.conf'
```
Comment (wagner-certat) [inline contrib/systemd/conf.py]: This is in intelmq.RUNTIME_CONF_FILE

File: contrib/systemd/conf.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+PIPELINE_CONF = INTELMQ_DIR+'/etc/pipeline.conf'
```
Comment (wagner-certat) [inline contrib/systemd/conf.py]: Is in `intelmq.PIPELINE_CONF_FILE`

Comment (navtej) [inline contrib/systemd/conf.py]: I did not imported intelmq, perhaps will be a good idea to do that.

**Comment by navtej** [comment]
This needs a lot more testing. I have tested it only on one OS. Will be great if more folks can test it out on other systems.

File: contrib/systemd/systemd.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                qb += 1
+                dsq_count[q] = qb
+            else:
+                dsq_count[q] = 1
+    return (sqs2bot, dsq_count)
+
+# Return only the bots which are directly connected
+# If a bot has more than one inputs or outputs stop processing
+#
+def connected_bots(bot,rc_data,pipe_data):
```
Comment (wagner-certat) [inline contrib/systemd/systemd.py]: That's not used

File: contrib/systemd/conf.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+#RUNTIME_CONF = INTELMQ_DIR+'/etc/runtime.conf'
+#PIPELINE_CONF = INTELMQ_DIR+'/etc/pipeline.conf'
+RUNTIME_CONF = intelmq.RUNTIME_CONF_FILE
+PIPELINE_CONF = intelmq.PIPELINE_CONF_FILE
+SYSTEMD_OUTPUT_DIR = INTELMQ_DIR+'/etc/systemd'
+SERVICE_PREFIX = "intelmq."
+DISABLE_IN_CONF = True
+SET_RUNMODE_IN_CONF = True
+INTELMQCTL_BIN = shutil.which('intelmqctl')
+SYSTEMCTL_BIN = shutil.which('systemctl')
```
Comment (wagner-certat) [inline contrib/systemd/conf.py]: never used

File: contrib/systemd/systemd.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                                               RANDOMIZE_DELAYS = RANDOMIZE_DELAYS,
+                                               ON_ACTIVE_SEC = ON_ACTIVE_SEC,
+                                               bot_interval = bot_interval,
+                                               bot_service_name = bot_service_name
+                                               )
+        with open(timer_file_name, "w", encoding='utf-8') as tmr_file:
+            tmr_file.write(timer_data)
+
+    if DISABLE_IN_CONF or SET_RUNMODE_IN_CONF:
+        shutil.move(RUNTIME_CONF, RUNTIME_CONF+'.bak')
```
Comment (wagner-certat) [inline contrib/systemd/systemd.py]: if you copy here, you do not need to fix the permissions afterwards

**Comment by wagner-certat** [comment]
I pushed a commit with a bunch of style fixes

File: contrib/systemd/systemd.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+                                               bot_service_name=bot_service_name
+                                               )
+        with open(timer_file_name, "w", encoding='utf-8') as tmr_file:
+            tmr_file.write(timer_data)
+
+    if DISABLE_IN_CONF or SET_RUNMODE_IN_CONF:
+        shutil.move(RUNTIME_CONF_FILE, RUNTIME_CONF_FILE + '.bak')
+        rc_data = collections.OrderedDict(sorted(rc_data.items()))
+        data = json.dumps(rc_data, indent=4)
+        with open(RUNTIME_CONF_FILE, "w", encoding='utf-8') as rc_file:
```
Comment (wagner-certat) [inline contrib/systemd/systemd.py]: You can reuse intelmqctl's [`write_updated_runtime_config`](https://github.com/certtools/intelmq/blob/bc272d1f7ca68107ad6e0313728c051b7cf76add/intelmq/bin/intelmqctl.py#L611) here.

**Comment by wagner-certat** [comment]
You are assuming that all collectors should be scheduled, but stream collectors (http stream, n6, xmpp, alienvault otx) can't run in this mode properly.

File: contrib/systemd/systemd.py (hunk 3 above)
Comment (navtej) [inline contrib/systemd/systemd.py]: I know. Not as of now, I am still trying to make the directly connected units (as explained in mailing list) work as connected systemd scripts. So can we please keep these? Unfortunately i cannot commit to any timeline, atleast this week looks pretty busy.

File: contrib/systemd/conf.py (hunk 4 above)
Comment (navtej) [inline contrib/systemd/conf.py]: SYSTEMCTL_BIN is the only one which is never used, however it was used for making other units trigger after the collector.

**Comment by navtej** [comment]
I understand, however , I am unable to find a good solution to it, couple of things I can think of
1. Let the user mark the bot's run_mode and only convert scheduled bots to services?
2. Take bot IDs as command line arguments, and convert those to services.

2 can then have web UI page to automate the work.
I might not be able to work on these for few days.

**Comment by navtej** [comment]
partially addresses the problem of selectively scheduling the bots. Unless there is a way to figure out which bots can be scheduled, there doesn't seems to be any decent way out of it.

**Comment by wagner-certat** [comment]
Could you reduce the number of tiny commits here (these "wip" and similar) please? Then I'll merge this and afterwards open a PR adding support for continuous collectors and some other fixes.

**Comment by wagner-certat** [comment]
Squashed the commits now on my own and merged it finally, also with some extensions and fixes by me (see https://github.com/certtools/intelmq/commit/83bef5542930a1d6dabae1ef73bfef72fd9d8513 https://github.com/certtools/intelmq/commit/c3ec3c684d204c31fcf0646eca7c0d64fc230afc and https://github.com/certtools/intelmq/commit/543d736a99140ee134152a5375707d6a4f1975b3)

**Comment by navtej** [comment]
Thanks. Sorry for be being too late, I have been pulled into too many things.

**Pull request closed without merging**